frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -1081,22 +1214,38 @@ class MostSpecificCandidate {
   friend class MostSpecificCandidates;
 
   const TypedFnSignature* fn_;
+  owned<FormalActualMap> faMap_;
```

Comment:
This seems fine provided that we aren't using `MostSpecificCandidate` / vectors / etc of it as query arguments or return values. If we are, we might be benefiting from reusing some queries & we won't be able to do that anymore with FormalActualMap here, because now it will depend on the call site. (That was true before with `constRefCoercionActual_` but that is much more minor than starting to include the actual types here). Or maybe we can null out `faMap_` in such cases?

So, I am wondering, are we using `MostSpecificCandidate` as a query result or argument type today?

I'm not worried about it being included in RRBPID which your PR message mentions. Are there other cases? `isCandidateApplicableInitialQuery` seems the closest to doing so that I can find in resolution-queries.cpp but it doesn't actually use MostSpecificCandidate(s).

Could you poke through other resolution queries (in other .cpp files in the resolution directory) to see if, indeed, `MostSpecificCandidate` is only used in RRBPID as a query argument/result?

Reply:
My intention was the PR and the `==` and `hash` changes was to allow `MostSpecificCandidate` etc. to work in queries as before. This is why I added a deep comparison to its `==` function, and specialized pointer hash to recurse into the pointer. You are right that this didn't work at the time of your review (I was missing `::update` and `::swap`). I added them in 17c618e43bd082e7f83730f895a04db5c6b3dce8, as well as a test to ensure that both `MostSpecificCandidate` and `MostSpecificCandidates` are appropriately compared and result in re-use in the query system.

Reply:
My question is about query reuse. If we add information about the call, we are making it less likely that a query working with a `MostSpecificCandidate` argument is going to be reused.

I think that the situation you had, where `MostSpecificCandidate` didn't actually work as a query argument, indicates that this is not an issue today, because presumably testing was passing in that mode. Right?

Reply:
Ah, I see what you're saying. Yes, testing was passing. I'm surprised to report that in basic experiments with `testInteractive`, the number of queries executed when re-running in a new generation hasn't changed with the addition of maps to candidates. So the effects seem minimal, though I agree that it would affect query re-use if, e.g., a file changes such that a function has a new formal, and a new actual (thus, the candidate is the same, but the FAMap no longer is).